lib/backends/sdc/containers.js

```diff
@@ -681,43 +667,96 @@ function defaultFabricNetwork(opts, container, payload, callback) {
+        log.debug({ results: networks, user: opts.account.uuid },
+            'Networks: Found %s results for query %s',
+            networks ? networks.length : 0, query);
+        callback(err, networks);
+        return;
+    }
+
+    // NOTE: the order of the functions in this parallel() call is significant;
+    // they are ordered by how docker prefers to resolve IDs:
+    // 1. exact id match
+    // 2. exact name match
+    // 3. partial id match
     vasync.parallel({
         funcs: [
+            function byExactId(cb) {
+                // length === 64, and 0..31 === 32..63, or it's an 'impossible'
+                // id under our double-uuid convention, and we can skip it.
+                if (query.substr(0, 32) !== query.substr(32)) {
+                    log.debug({ query: query },
+                        'Networks: impossible exactId: %s, skipping', query);
+                    setImmediate(cb, null, []);
+                    return;
+                }
+                var uuid = utils.shortNetworkIdToUuid(query);
+
+                // XXX - ldapEscape required to work around NAPI-367
+                var listParams = {
+                    uuid: utils.ldapEscape(uuid),
+                    fabric: true,
+                    owner_uuid: opts.account.uuid
+                };
+
+                napi.listNetworks(listParams,
+                    { headers: { 'x-request-id': opts.req_id }}, cb);
+            },
             function byName(cb) {
                 var listParams = {
                     name: query,
                     fabric: true,
-                    owner_uuid: opts.account.uuid,
-                    limit: 2
+                    owner_uuid: opts.account.uuid
                 };
 
                 log.debug({ listParams: listParams },
                     format('Networks: searching for network %s',
                         listParams.name));
 
-
                 napi.listNetworks(listParams,
                     { headers: {'x-request-id': opts.req_id }}, cb);
             },
             function byDockerId(cb) {
-                // to search for docker ids, when NAPI only retains UUIDs, we
-                // rely on the 'double uuid' convention (see
-                // utils.networkUuidToDockerId()). To perform the search,
-                // we truncate the provided string to 8 chars, perform a
-                // wildcard search on networks, and post-filter the results to
-                // find more-complete matches than the at-most-8-char prefix
-                // might find.
+                // we assume the 'double uuid' convention for networks here,
+                // that is, dockerId = (uuid + uuid).replace(/-/g, '').
+                // So if we have a query.length > 31, 32... must be a prefix of
+                // 0..31. If not, the id supplied is impossible in our system.
+                if (query.length > 31
+                    && query.substr(0, query.substr(32).length)
```

Comment:
if

```
query.substr(0, query.substr(32).length)
```

is used to get the first 32 characters of the string `query`, would

```
query.substr(0, 32)
```

be an equivalent but simpler way to achieve that?